pkg/fqdn/cache.go

```diff
@@ -624,6 +624,19 @@ func (c *DNSCache) Dump() (lookups []*cacheEntry) {
 	return deduped
 }
 
+// Count returns the count of the still-valid entries in the DNS cache. Note,
+// this count is not deduplicated, see Dump().
```

Comment:
Question: why is this the correct count for this metric? Why not just the total number of entries?

Reply:
Or rather to re-word my question, why start the count at `uint64(len(c.forward))`?

Reply:
Good question. It depends on what you mean by "total number of entries" and which entries. 

If we take a look at the `DNSCache` struct:

```go
type DNSCache struct {
	lock.RWMutex

	// forward DNS lookups name -> IPEntries
	// IPEntries maps IP -> entry that provides it. An entry may provide multiple IPs.
	forward map[string]ipEntries

	// IP->dnsNames lookup
	// This map is subordinate to forward, above. An IP inserted into forward, or
	// expired in forward, should also be added/removed in reverse.
	reverse map[string]nameEntries

	// LastCleanup is the latest time for which entries have been expired. It is
	// used as "now" when doing lookups and advanced by calls to .GC
	// When an entry is added with an expiration time before lastCleanup, it is
	// set to that value.
	lastCleanup time.Time

	// cleanup maps the TTL expiration times (in seconds since the epoch) to
	// DNS names that expire in that second. On every new insertion where the
	// new data is actually inserted into the cache (i.e. it expires later than
	// an existing entry) cleanup will be updated. CleanupExpiredEntries cleans
	// up these entries on demand.
	// Note: Lookup functions will not return expired entries, and this is used
	// to proactively enforce expirations.
	// Note: It is important to periodically call CleanupExpiredEntries
	// otherwise this map will grow forever.
	cleanup map[int64][]string

	// overLimit is a set of DNS names that were over the per-host configured
	// limit when they received an update. The excess IPs will be removed when
	// cleanupOverLimitEntries is called, but will continue to be returned by
	// Lookup until then.
	// Note: It is important to periodically call GC otherwise this map will
	// grow forever (it is very bounded, however).
	overLimit map[string]bool

	// perHostLimit is the number of maximum number of IP per host.
	perHostLimit int

	// minTTL is the minimun TTL value that a cache entry can have, if the TTL
	// sent in the Update is lower, the TTL will be owerwritten to this value.
	// Due is only read-only is not protected by the mutex.
	minTTL int
}
```

We see that `cleanup` and `overLimit` clearly not of interest based on the godoc. If we look at `reverse`'s godoc, we see that it's the same as `forward` but with the key-val inverted. 

`forward` is clearly a map from FQDN name (`google.com`) to many IPs. We know that when a Cilium-managed endpoint makes a DNS request (while the endpoint is selected by a [DNS rule](https://docs.cilium.io/en/latest/gettingstarted/dns/#dns-policies-using-patterns)), then these requests are cached inside `DNSCache` via https://github.com/cilium/cilium/blob/636edd09da749f5fbeea2ae7c5eaa2e97d4d682e/daemon/cmd/fqdn.go#L511.

Hence we are `Count()`ing the number of entries from `forward`.

Hopefully this answers your question. Happy to dig further as well.

Reply:
Good catch @tommyp1ckles. This is indeed incorrect. I've split the `Count()` to return two return values: one for names and one for IPs. 

So in your above example we'd get 1 name and 2 IPs.